### ISSUE TITLE
child bundle apps not deleted when parent app (dynamic profile) and root app (cluster) deleted 

### ISSUE THREAD

**Issue by bcle** [platform9/arlon#17]
When a top-level arlon Cluster application is deleted in argocd, the child dynamic profile app (if the profile is dynamic) is deleted as expected, but the profile's child apps (one for each bundle) are not. This appears to be an ArgoCD bug, and it appears to affect application resources that appear under the `templates/` subdirectory of a Helm chart.

Experiment:
- Keep an empty arlon cluster "cluster-a" by deleting (in git and manually as needed) all children apps.
- Manually create a parent app (using `argocd app create`) that has the same definition as the equivalent dynamic profile used in the bug scenario. The actual command I used is: `argocd app create standalone-dynprofile --repo https://github.com/bcle/fleet-infra.git --path profiles/dyntest-1/mgmt --dest-namespace argocd --dest-server https://kubernetes.default.svc --auto-prune --sync-option Prune=true --sync-policy automated --helm-set clusterName=cluster-a`
- Observe the 2 child apps being created and successfully synced
- Manually delete the parent app: `argocd app delete -y standalone-dynprofile`
- Observe that the 2 child apps are automatically deleted as expected
- Repeat the experiment, but this time by checking into git a `profile.yaml` file under the cluster's root app's `templates/` directory
- Observe the creation of the child apps
- In git, remove `profile.yaml` and push the change
- Observe that the parent app is deleted, but not the child apps

**Comment by bcle** [comment]
Another experiment: if I add 2 individual child apps directly to the root cluster's app's `templates` directory in git, and later revert the change in git, the child apps are created and later automatically deleted as expected. So the bug seems to occur when there's an intermediate app (the dynamic profile app) involved.

**Comment by bcle** [comment]
And another experiment, which suggests that this bug is not specific to Helm charts:
- In git, I manually created a top-level cluster app for `cluster-b`. The manifest format is a single, plain YAML file that I generated from `cluster-a`'s chart using `helm template`. I made sure all occurences of `cluster-a` were renamed to `cluster-b`. I also made sure the combined YAML (named `everything.yaml`) included an application resource for `profile-dyntest-1`.
- I manually created the top-level app by executing `argocd app create cluster-b --repo https://github.com/bcle/fleet-infra.git --path clusters/cluster-b --dest-namespace argocd --dest-server https://kubernetes.default.svc --auto-prune --sync-option Prune=true --sync-policy automated`
- I confirmed that `cluster-b`, `cluster-b-profile-dyntest-1`, `cluster-b-xenial`, and `cluster-b-guestbook-ref` were all created and eventually healthy as expected
- In git, I edited `everything.yaml` to remove the `profile-dyntest-1` application resource.
- After pushing the commit, in argocd I observed that `cluster-b-profile-dyntest-1` is gone, but not the two child apps. They're still there.
- In conclusion, the bug seems to occur in the presence of an intermediate app resource (the dynamic profile) which is included in either a plain YAML manifest or Helm Chart.
- As shown in an earlier experiment, the bug does NOT occur if the intermediate app resource is standalone (not part of another manifest).

**Comment by bcle** [comment]
More experimentation:
- Manually edited root app resource to set auto-sync / prune
- Used argocd CLI to delete `cluster-b-profile-dyntest-1`. This causes child apps to be deleted as well.
- Restored the profile app to repeat the experiment
- This time, deleted the profile app using kubectl: `k -n argocd delete application cluster-b-profile-dyntest-1`. The child apps are NOT deleted.
- Conclusion: the behavior of deleting the app resource directly in k8s (via kubectl) is different from doing it using argocd API. So the API does something special and knows to delete an app's child apps.

**Comment by bcle** [comment]
Good news: breakthrough discovery: when using argocd CLI to delete an app via the API, "cascade" deletion is enabled by default. In the server, this is implemented by adding a finalizer to the resource before deleting it from Kubernetes. This results in proper deletion of child applications. Experiment: I manually added `resources-finalizer.argocd.argoproj.io/foreground` to the dynamic profile's application resource, and deleting the resource via `kubectl` has the intended cascade effect. We have a solution. Code fix is being tested.

**Comment by cruizen** [comment]
Hi @bcle , is this issue resolved with your last commit?
